# SPDX-FileCopyrightText: 2002-2025 Lawrence Livermore National Laboratory & other contributors to SLURM <https://github.com/SchedMD/slurm/commits/master/DISCLAIMER>
# SPDX-License-Identifier: GPL-2.0-or-later
# SPDX-PackageName: slurm-containers
# Backport of upstream 26ece4554f "Add closeall_except()." (Issue 50093)
commit 26ece4554fccb2043f0af45b635f2929caa204f1
Author: Nathan Rini <nate@schedmd.com>
Date:   Tue Oct 22 11:24:29 2024 -0600

    Add closeall_except().
    
    Expand closeall() to allow an array of file descriptors to skip over
    instead of closing.
    
    Issue 50093

diff --git a/src/common/fd.c b/src/common/fd.c
index 9393162079..57ca623409 100644
--- a/src/common/fd.c
+++ b/src/common/fd.c
@@ -89,6 +89,7 @@ do { \
  * for details.
  */
 strong_alias(closeall, slurm_closeall);
+strong_alias(closeall_except, slurm_closeall_except);
 strong_alias(fd_close, slurm_fd_close);
 strong_alias(fd_set_blocking,	slurm_fd_set_blocking);
 strong_alias(fd_set_nonblocking,slurm_fd_set_nonblocking);
@@ -100,7 +101,19 @@ strong_alias(rmdir_recursive, slurm_rmdir_recursive);
 static int fd_get_lock(int fd, int cmd, int type);
 static pid_t fd_test_lock(int fd, int type);
 
-static void _slow_closeall(int fd)
+static bool _is_fd_skipped(int fd, int *skipped)
+{
+	if (!skipped)
+		return false;
+
+	for (int i = 0; skipped[i] >= 0; i++)
+		if (fd == skipped[i])
+			return true;
+
+	return false;
+}
+
+static void _slow_closeall(int fd, int *skipped)
 {
 	struct rlimit rlim;
 
@@ -109,11 +122,12 @@ static void _slow_closeall(int fd)
 		rlim.rlim_cur = 4096;
 	}
 
-	while (fd < rlim.rlim_cur)
-		close(fd++);
+	for (; fd < rlim.rlim_cur; fd++)
+		if (!_is_fd_skipped(fd, skipped))
+			close(fd);
 }
 
-extern void closeall(int fd)
+extern void closeall_except(int fd, int *skipped)
 {
 	char *name = "/proc/self/fd";
 	DIR *d;
@@ -128,7 +142,7 @@ extern void closeall(int fd)
 	if (!(d = opendir(name))) {
 		debug("Could not read open files from %s: %m, closing all potential file descriptors",
 		      name);
-		_slow_closeall(fd);
+		_slow_closeall(fd, skipped);
 		return;
 	}
 
@@ -136,13 +150,20 @@ extern void closeall(int fd)
 		/* Ignore "." and ".." entries */
 		if (dir->d_type != DT_DIR) {
 			int open_fd = atoi(dir->d_name);
-			if (open_fd >= fd)
+
+			if ((open_fd >= fd) &&
+			    !_is_fd_skipped(open_fd, skipped))
 				close(open_fd);
 		}
 	}
 	closedir(d);
 }
 
+extern void closeall(int fd)
+{
+	return closeall_except(fd, NULL);
+}
+
 extern void fd_close(int *fd)
 {
 	if (fd && *fd >= 0) {
diff --git a/src/common/fd.h b/src/common/fd.h
index 18bd277262..510aa39ce3 100644
--- a/src/common/fd.h
+++ b/src/common/fd.h
@@ -49,6 +49,14 @@
 /* close all FDs >= a specified value */
 extern void closeall(int fd);
 
+/*
+ * close all FDs >= a specified value except FDs in skipped array
+ * IN fd - start closing file descriptors at fd
+ * IN skipped - array of file descriptors to skip (or NULL)
+ *              array must be terminated by -1 if provided
+ */
+extern void closeall_except(int fd, int *skipped);
+
 /* Close a specific file descriptor and replace it with -1 */
 extern void fd_close(int *fd);
 
diff --git a/src/common/slurm_xlator.h b/src/common/slurm_xlator.h
index 92425b3545..4aed2bd1a5 100644
--- a/src/common/slurm_xlator.h
+++ b/src/common/slurm_xlator.h
@@ -115,6 +115,7 @@
 
 /* fd.[ch] functions */
 #define closeall		slurm_closeall
+#define closeall_except		slurm_closeall_except
 #define fd_set_blocking		slurm_fd_set_blocking
 #define fd_set_nonblocking	slurm_fd_set_nonblocking
 #define fd_get_socket_error	slurm_fd_get_socket_error
